<a href="https://colab.research.google.com/github/Ccuevas3410/Artificial-Intelligence/blob/master/Assignment%204.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


In [44]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.model_selection import train_test_split

In [45]:
# Load the dataset (load remotely, not locally)
irisData = load_iris()
# Output the first 15 rows of the data
irisDf = pd.DataFrame(data= np.c_[irisData['data'], irisData['target']], columns= irisData['feature_names'] + ['target'])
# Display a summary of the table information (number of datapoints, etc.)
irisDf.head(15)


sepal length (cm)  sepal width (cm)  ...  petal width (cm)  target
0                 5.1               3.5  ...               0.2     0.0
1                 4.9               3.0  ...               0.2     0.0
2                 4.7               3.2  ...               0.2     0.0
3                 4.6               3.1  ...               0.2     0.0
4                 5.0               3.6  ...               0.2     0.0
5                 5.4               3.9  ...               0.4     0.0
6                 4.6               3.4  ...               0.3     0.0
7                 5.0               3.4  ...               0.2     0.0
8                 4.4               2.9  ...               0.2     0.0
9                 4.9               3.1  ...               0.1     0.0
10                5.4               3.7  ...               0.2     0.0
11                4.8               3.4  ...               0.2     0.0
12                4.8               3.0  ...               0.1     0.0
13                4.3               3.0  ...               0.1     0.0
14                5.8               4.0  ...               0.2     0.0

[15 rows x 5 columns]

In [46]:
irisDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    float64
dtypes: float64(5)
memory usage: 6.0 KB


# About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data is made of 150 rows with non-null values and 5 columns which are:

Features: Sepal length, Sepal width, Petal length, Petal width
Label: target

# Part 2: Split the dataset into train and test

In [47]:
# Take the dataset and split it into our features (X) and label (y)
x = irisData.data
y = irisData.target
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(x, y, train_size=.90, random_state=40)

# Part 3: Logistic Regression

In [48]:
# i. Use sklearn to train a LogisticRegression model on the training set

regression = LogisticRegression()
regression.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

y_pred = regression.predict(X_test)
y_pred_prob = regression.predict_proba(X_test)
posibbleClasses = load_iris().target_names
for class_name, probability in zip(posibbleClasses, y_pred_prob):
    index = np.argmax(probability)
    print("predicted class: " + str(class_name) +   " confidence: " + str(probability[index]*100))
    
# iii. Report on the score for Logistic regression model, what does the score measure?
score = regression.score(X_test, y_test)
print("Score is", score)

confusion_matrix(y_test,regression.predict(X_test))


predicted class: setosa confidence: 98.49952418172454
predicted class: versicolor confidence: 74.04623300557905
predicted class: virginica confidence: 93.85834949832753
Score is 1.0


array([[4, 0, 0],
       [0, 7, 0],
       [0, 0, 4]])

We can see that the score for this model stands on 100%. <br>
The confusion matrix showes setosa we have 4 correct predicted setosa and 0 false valuse. Conclusion: very high accuracy. <br>
For versicolor we got 7 predicted correclty and 0 for falsely predictions (all are zeros). <br>
For virginicia again we got 4 correct predictions with no false predictions, which again tells up we have very high accuracy. 


In [49]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
intercepts = regression.intercept_
coeff = regression.coef_.T
print("intercepts: ", intercepts)
print( "coeff:" , coeff)

intercepts:  [  9.28765789   1.71996682 -11.0076247 ]
coeff: [[-0.3800456   0.59261374 -0.21256814]
 [ 0.96204981 -0.30949256 -0.65255725]
 [-2.4418961  -0.22740222  2.66929832]
 [-1.05252963 -0.85501325  1.90754288]]


# Part 4: Support Vector Machine

In [50]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc = svm.SVC(probability=True)
svc.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

y_pred_svc = svc.predict(X_test)
y_pred_prob_svc = svc.predict_proba(X_test)
for class_name, proba in zip(posibbleClasses, y_pred_prob_svc):
    index = np.argmax(proba)
    print("Predicted class :", class_name ,"\nConfidence :", proba[index])

# iii. Report on the score for the SVM, what does the score measure?

score = svc.score(X_test, y_test)
print("Socre:", score)

print(confusion_matrix(y_test,svc.predict(X_test)))


Predicted class : setosa 
Confidence : 0.9734339875025979
Predicted class : versicolor 
Confidence : 0.8388566195319529
Predicted class : virginica 
Confidence : 0.9780935498204312
Socre: 1.0
[[4 0 0]
 [0 7 0]
 [0 0 4]]


Once again, we see that the coeff matrix produces high accuracy, as the predicted vs. the actual classes match with zeros false predictions accordingly.  <br> Score here also stands on 100%.

# Part 5: Neural Network

In [51]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier(solver='adam',learning_rate_init=0.01, hidden_layer_sizes=10, max_iter=600,random_state=1)
mlp.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_predict_mlp = mlp.predict(X_test)
y_pred_prob_mlp = mlp.predict_proba(X_test)

for class_name, probability in zip(posibbleClasses, y_pred_prob_mlp):
    index = np.argmax(probability)
    print("Predicted class: ", class_name ,", Confidence:" ,probability[index])

    # iii. Report on the score for the Neural Network, what does the score measure?
score = mlp.score(X_test, y_test)
print("Accuracy score: ",score)

# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

Predicted class:  setosa , Confidence: 0.9996544449712463
Predicted class:  versicolor , Confidence: 0.8836402049279067
Predicted class:  virginica , Confidence: 0.9962550196137152
Accuracy score:  1.0


For this model the score here produces 93% percision.

# Part 6: K-Nearest Neighbors

In [52]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

KNN = KNeighborsClassifier(n_neighbors=8)
KNN.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

y_pred_knn = KNN.predict(X_test)
y_pred_prob_knn = KNN.predict_proba(X_test)
for class_name, probability in zip(posibbleClasses, y_pred_prob_knn):
    index = np.argmax(probability)
    print("predicted class:", class_name,", confidence:", probability[index])
# iii. Report on the score for kNN, what does the score measure?
    score = KNN.score(X_test, y_test)


print("Accuracy: ",score)

predicted class: setosa , confidence: 1.0
predicted class: versicolor , confidence: 1.0
predicted class: virginica , confidence: 1.0
Accuracy:  1.0


For this model the score here produces 100% percision.

# Part 7: Conclusions and takeaways


In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

Answer:

In this assignment we used 4 classifiers. Logistic Regression
Support Vector Machine, Neural Network, K-Nearest Neighbors

All models except the neural network classifier gave a pretty accurate result, i
I am surprised

